File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode result = null;
        if(head == null || head.next == null)
            return result;
        ListNode slow = head;
        ListNode fast = head;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
            
            if(slow == fast){
                slow = head;
                while(slow != fast){
                    slow = slow.next;
                    fast = fast.next;
                }
                result = fast;
                return result;
            }
        }
        return result;
    }
}
// HashMap<ListNode,Integer> hp = new HashMap<>();
//         ListNode temp = head;
//         int i = 0,pos = -1;
//         while(temp != null){
//             if(hp.containsKey(temp)){
//                 pos = hp.get(temp);
//                 break;
//             }else{
//                 hp.put(temp,i);
//                 i++;
//             }
//             temp = temp.next;
//         }
//         if(pos != -1){
//             temp = head;
//             i = 0;
//             while(i <= pos && temp != null){
//                 if(i == pos){
//                     return temp;
//                 }
//                 temp = temp.next;
//                 pos--;
//             }
//         }
//         return null;

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
    if head == nil || head.Next == nil{
        return nil
    }
    slow := head.Next
    fast := head.Next.Next
    for fast != nil && fast.Next != nil{
        if slow == fast{
            slow = head
            for(slow != fast){
                fast = fast.Next
                slow = slow.Next
            }
            return slow
        }
        slow = slow.Next
        fast = fast.Next.Next
    }
    return nil
}